tests/vm_state_test.go

```diff
@@ -0,0 +1,216 @@
+				&expect.BSnd{S: "tpm2_createprimary -Q --hierarchy=o --key-context=prim.ctx\n"},
+				&expect.BExp{R: console.PromptExpression},
+				&expect.BSnd{S: "echo MYSECRET | tpm2_create --hash-algorithm=sha256 --public=seal.pub --private=seal.priv --sealing-input=- --parent-context=prim.ctx\n"},
+				&expect.BExp{R: console.PromptExpression},
+				&expect.BSnd{S: "tpm2_load -Q --parent-context=prim.ctx --public=seal.pub --private=seal.priv --name=seal.name --key-context=seal.ctx\n"},
+				&expect.BExp{R: console.PromptExpression},
+				&expect.BSnd{S: "tpm2_evictcontrol --hierarchy=o --object-context=seal.ctx 0x81010002\n"},
+				&expect.BExp{R: console.PromptExpression},
+				&expect.BSnd{S: "tpm2_unseal -Q --object-context=0x81010002\n"},
+				&expect.BExp{R: "MYSECRET"},
+			}, 300)).To(Succeed(), "failed to store secret into the TPM")
+		}
+
+		checkTPM := func(vmi *v1.VirtualMachineInstance) {
+			By("Ensuring the TPM is still functional and its state carried over")
+			ExpectWithOffset(1, console.SafeExpectBatch(vmi, []expect.Batcher{
+				&expect.BSnd{S: "tpm2_unseal -Q --object-context=0x81010002\n"},
+				&expect.BExp{R: "MYSECRET"},
+			}, 300)).To(Succeed(), "the state of the TPM did not persist")
+		}
+
+		addDataToEFI := func(vmi *v1.VirtualMachineInstance) {
+			By("Creating an efivar")
+			cmd := `printf "\x07\x00\x00\x00\x42" > /sys/firmware/efi/efivars/kvtest-12345678-1234-1234-1234-123456789abc`
+			err = console.RunCommand(vmi, cmd, 10*time.Second)
+			Expect(err).NotTo(HaveOccurred())
+		}
+
+		checkEFI := func(vmi *v1.VirtualMachineInstance) {
+			By("Ensuring the efivar is present")
+			Expect(console.SafeExpectBatch(vmi, []expect.Batcher{
+				&expect.BSnd{S: "hexdump /sys/firmware/efi/efivars/kvtest-12345678-1234-1234-1234-123456789abc\n"},
+				&expect.BExp{R: "0042"},
+			}, 10)).To(Succeed(), "expected efivar is missing")
+		}
+
+		DescribeTable("should persist VM state of", func(withTPM, withEFI bool, ops ...string) {
+			By("Creating a migratable Fedora VM with UEFI")
+			vmi := libvmi.NewFedora(
+				libvmi.WithInterface(libvmi.InterfaceDeviceWithMasqueradeBinding()),
+				libvmi.WithNetwork(v1.DefaultPodNetwork()),
+				libvmi.WithCloudInitNoCloudNetworkData(libnet.CreateDefaultCloudInitNetworkData()),
+				libvmi.WithUefi(false),
+				libvmi.WithResourceMemory("1Gi"),
+			)
+			vmi.Namespace = util.NamespaceTestDefault
+			if withTPM {
+				By("with persistent TPM enabled")
+				vmi.Spec.Domain.Devices.TPM = &v1.TPMDevice{
+					Persistent: pointer.BoolPtr(true),
+				}
+			}
+			if withEFI {
+				By("with persistent EFI enabled")
+				vmi.Spec.Domain.Firmware = &v1.Firmware{
+					Bootloader: &v1.Bootloader{
+						EFI: &v1.EFI{SecureBoot: pointer.BoolPtr(false), Persistent: pointer.BoolPtr(true)},
+					},
+				}
+			}
+			vm := tests.NewRandomVirtualMachine(vmi, false)
+			vm, err = virtClient.VirtualMachine(util.NamespaceTestDefault).Create(context.Background(), vm)
```

Comment:
It's kind of a nit but it's usually better to use `testsuite.GetTestNamespace(vmi)` here, otherwise if this test is run on a cluster configured to use root VMs it's going to fail.